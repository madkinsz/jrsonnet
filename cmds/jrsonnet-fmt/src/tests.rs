use dprint_core::formatting::PrintOptions;
use indoc::indoc;

use crate::Printable;

fn reformat(input: &str) -> String {
	let (source, _) = jrsonnet_rowan_parser::parse(input);

	dprint_core::formatting::format(
		|| source.print(),
		PrintOptions {
			indent_width: 2,
			max_width: 100,
			use_tabs: false,
			new_line_text: "\n",
		},
	)
}

macro_rules! assert_formatted {
	($input:literal, $output:literal) => {
		let formatted = reformat(indoc!($input));
		let mut expected = indoc!($output).to_owned();
		expected.push('\n');
		if formatted != expected {
			panic!(
				"bad formatting, expected\n```\n{formatted}\n```\nto be equal to\n```\n{expected}\n```",
			)
		}
	};
}

#[test]
fn padding_stripped_for_multiline_comment() {
	assert_formatted!(
		"{
            /*
                Hello
                    World
            */
            _: null,
        }",
		"{
          /*
          Hello
              World
          */
          _: null,
        }"
	);
}

#[test]
fn last_comment_respects_spacing_with_inline_comment_above() {
	assert_formatted!(
		"{
			a: '', // Inline

			// Comment
        }",
		"{
		  a: '', // Inline

		  // Comment
		}"
	);
}

#[test]
fn complex_comments_snapshot() {
	insta::assert_display_snapshot!(reformat(indoc!(
		"{
		  comments: {
			_: '',
			//     Plain comment
			a: '',

			#    Plain comment with empty line before
			b: '',
			/*Single-line multiline comment

			*/
			c: '',

			/**Single-line multiline doc comment

			*/
			c: '',

			/**Multiline doc
			Comment
			*/
			c: '',

			/*

	Multi-line

	comment
			*/
			d: '',

			e: '', // Inline comment

			k: '',

			// Text after everything
		  },
		  comments2: {
			k: '',
			// Text after everything, but no newline above
		  },
          spacing: {
            a: '',

            b: '',
          },
          noSpacing: {
            a: '',
            b: '',
          },
        }"
	)))
}
