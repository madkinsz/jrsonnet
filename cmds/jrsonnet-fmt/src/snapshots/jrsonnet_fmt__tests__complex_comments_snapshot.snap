---
source: cmds/jrsonnet-fmt/src/tests.rs
expression: "reformat(indoc!(\"{\n\t\t  comments: {\n\t\t\t_: '',\n\t\t\t//     Plain comment\n\t\t\ta: '',\n\n\t\t\t#    Plain comment with empty line before\n\t\t\tb: '',\n\t\t\t/*Single-line multiline comment\n\n\t\t\t*/\n\t\t\tc: '',\n\n\t\t\t/**Single-line multiline doc comment\n\n\t\t\t*/\n\t\t\tc: '',\n\n\t\t\t/**Multiline doc\n\t\t\tComment\n\t\t\t*/\n\t\t\tc: '',\n\n\t\t\t/*\n\n\tMulti-line\n\n\tcomment\n\t\t\t*/\n\t\t\td: '',\n\n\t\t\te: '', // Inline comment\n\n\t\t\tk: '',\n\n\t\t\t// Text after everything\n\t\t  },\n\t\t  comments2: {\n\t\t\tk: '',\n\t\t\t// Text after everything, but no newline above\n\t\t  },\n          spacing: {\n            a: '',\n\n            b: '',\n          },\n          noSpacing: {\n            a: '',\n            b: '',\n          },\n        }\"))"
---
{
  comments: {
    _: '',
    // Plain comment
    a: '',

    # Plain comment with empty line before
    b: '',
    /* Single-line multiline comment */
    c: '',

    /**
     * Single-line multiline doc comment
     */
    c: '',

    /**
     * Multiline doc
     * Comment
     */
    c: '',

    /*
    Multi-line

    comment
    */
    d: '',

    e: '', // Inline comment

    k: '',

    // Text after everything
  },
  comments2: {
    k: '',
    // Text after everything, but no newline above
  },
  spacing: {
    a: '',

    b: '',
  },
  noSpacing: {
    a: '',
    b: '',
  },
}
